packages/components/src/presentation-components/properties/inputs/ItemsLoader.tsx

```diff
@@ -0,0 +1,93 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+/** @internal */
+export const VALUE_BATCH_SIZE = 100;
+
+/** @internal */
+export const FILTER_WARNING_OPTION = { label: "Too many values please use filter", value: "__filter__", disabled: true };
+
+interface LoadedItems<T> {
+  options: T[];
+  length: number;
+  hasMore: boolean;
+}
+
+/** @internal */
+export class ItemsLoader<T> {
+  private _loadedItems: T[] = [];
+  private _isLoading = false;
+  private _hasMore = true;
+  private _disposed = false;
+  private _offset = 0;
+
+  constructor(
+    private _beforeLoad: () => void,
+    private _onItemsLoaded: (newItems: T[]) => void,
+    private _loadItems: (offSet: number) => Promise<LoadedItems<T>>,
+    private _getOptionLabel: (option: T) => string,
+  ) {}
+
+  public dispose() {
+    this._disposed = true;
+  }
+
+  public async loadMatchingItems(valuesToMatch?: string[]) {
+    const needsItemsLoaded = (items: T[]) => {
+      if (valuesToMatch && valuesToMatch.length > 0) {
+        const matchingItems = items.filter((item) => {
+          return valuesToMatch.some((valueToMatch) => valueToMatch === this._getOptionLabel(item));
+        });
+        return matchingItems.length < valuesToMatch.length;
+      }
+      return items.length < VALUE_BATCH_SIZE;
+    };
+
+    await this.loadUniqueItems(needsItemsLoaded);
+  }
+
+  public async loadItems(filterText?: string) {
+    const needsItemsLoaded = (options: T[]): boolean => {
+      if (!filterText) {
+        return options.length === 0;
```

Comment:
```suggestion
        return options.length < VALUE_BATCH_SIZE;
```

Reply:
If there are already loaded values we probably don't want to load more, because it already loaded up to the VALUE_BATCH_SIZE or there are not enough items to load.

Reply:
If there are not enough items to load this won't be called as `ItemsLoader` is tracking if there are more items to load. This changes does not change behavior. In my opinion it just makes both cases (wiht\without filterTest) the same - load items if there are less than batch size.